File: min_falling_path_sum.py
# 931. Minimum Falling Path Sum

"""
given an 'n  n' array of integers 'matrix', return the minimum sum of any
falling path through 'matrix'. a falling path starts at any element in the
first row and choose the element in the next row that is either directly
below or diagonally left/right. specifically, the next element from position
'(row, col)' will be '(row + 1, col - 1), (row + 1, col), or (row + 1, col +
1)'.
"""


class Solution(object):
    def minFallingPathSum(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        grid = matrix
        for i in range(1, len(grid)):
            for j in range(len(grid[i])):
                if i - 1 >= 0 and j - 1 >= 0:
                    p1 = grid[i - 1][j - 1]
                else:
                    p1 = float("inf")
                if i - 1 >= 0:
                    p2 = grid[i - 1][j]
                else:
                    p2 = float("inf")
                if i - 1 >= 0 and j + 1 <= len(grid[0]) - 1:
                    p3 = grid[i - 1][j + 1]
                else:
                    p3 = float("inf")
                grid[i][j] += min(p1, p2, p3)
        return min(grid[len(grid) - 1])


if __name__ == "__main__":
    obj = Solution()
    print(obj.minFallingPathSum(matrix=[[2, 1, 3], [6, 5, 4], [7, 8, 9]]))
    print(obj.minFallingPathSum(matrix=[[-19, 57], [-40, -5]]))

File: min_falling_path_sum.cpp
#include "leetcode.h"

/*
 * given NxN array of integers 'matrix' return the minimum
 * sum of any falling path through 'matrix'
 *
 * a falling path starts at any element in row 0 and chooses
 * the element in the next row that is either directly below
 * or diagonally left/right that obviously makes for the lowest sum
 */

class Solution {
public:
  int minFallingPathSum(vvd(int) & matrix) {
    vvd(int) dp(matrix.size(), vector<int>(matrix.size(), INT_MAX));
    for (int i = 0; i < matrix.size(); i++)
      dp[0][i] = matrix[0][i];
    for (int i = 1; i < matrix.size(); i++) {
      for (int j = 0; j < matrix.size(); j++) {
        if (j > 0 && j < matrix.size() - 1) {
          dp[i][j] =
              min(dp[i - 1][j - 1] + matrix[i][j], dp[i - 1][j] + matrix[i][j]);
          dp[i][j] = min(dp[i - 1][i + 1] + matrix[i][j], dp[i][j]);
        } else if (j == 0)
          dp[i][j] =
              min(dp[i - 1][j + 1] + matrix[i][j], dp[i - 1][j] + matrix[i][j]);
        else if (j == matrix.size() - 1)
          dp[i][j] =
              min(dp[i - 1][j - 1] + matrix[i][j], dp[i - 1][j] + matrix[i][j]);
      }
    }
    int ans = INT_MAX;
    for (auto &x : dp[dp.size() - 1])
      ans = min(ans, x);
    return ans;
  }
};

int main() {
  Solution obj;
  vvd(int) matrix = {{2, 1, 3},          // 1
                     {6, 5, 4},          //-> 4
                     {7, 8, 9}};         // 8 <-
  cout << obj.minFallingPathSum(matrix); // expect 13
}
